File: src/BehaviorDiff.Go/internal/rewriter/runtime.go
package rewriter

import (
	"fmt"
	"go/format"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

const runtimeSource = `package behaviordiffrt

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strings"
	"sync"
)

const (
	schema = "behaviordiff.trace/1"
	language = "go"
	writerCapacity = 1
)

type Member struct {
	Module string
	Method string
	File string
	Line int
	ReturnKind string
	SourceResolution string
	Status string
	SkipReason string
	Detail string
	IsTestRoot bool
	IsHarness bool
	ParameterNames []string
}

type capture struct {
	digest string
	rendered string
}

type Frame struct {
	member Member
	testID string
	callID uint64
	parentCallID uint64
	depth int
	ordinal uint64
	args *capture
	exit sync.Once
}

type moduleState struct {
	members map[string]Member
	tracedCalls uint64
}

type ordinalKey struct {
	testID string
	method string
}

var state = struct {
	sync.Mutex
	nextCallID uint64
	ordinals map[ordinalKey]uint64
	modules map[string]*moduleState
	trace *os.File
	tracePath string
	manifestPath string
	enqueued uint64
	written uint64
	digest Counters
}{
	nextCallID: 1,
	ordinals: make(map[ordinalKey]uint64),
	modules: make(map[string]*moduleState),
}

var defaultSensitiveNames = []string{"password", "token", "secret", "key", "ssn", "email", "auth", "credential"}
var sensitiveNames = append(defaultSensitiveNames, configuredList("BEHAVIORDIFF_REDACT_NAMES")...)
var digestOnlyTypes = configuredList("BEHAVIORDIFF_REDACT_TYPES")
var digestOnlyPaths = configuredList("BEHAVIORDIFF_REDACT_PATHS")

func init() {
	configured := os.Getenv("BEHAVIORDIFF_TRACE")
	if configured == "" {
		return
	}
	state.tracePath = decorate(configured, "")
	state.manifestPath = decorate(configured, ".manifest")
	if err := os.MkdirAll(filepath.Dir(state.tracePath), 0755); err != nil {
		panic(fmt.Errorf("behaviordiff: create trace directory: %w", err))
	}
	trace, err := os.Create(state.tracePath)
	if err != nil {
		panic(fmt.Errorf("behaviordiff: create trace: %w", err))
	}
	state.trace = trace
	state.Lock()
	defer state.Unlock()
	writeManifestLocked()
}

func Register(members ...Member) bool {
	state.Lock()
	defer state.Unlock()
	changed := false
	for _, member := range members {
		changed = ensureMemberLocked(member) || changed
	}
	if changed {
		writeManifestLocked()
	}
	return true
}

func TypeName[T any]() string {
	return stableTypeName(reflect.TypeOf((*T)(nil)).Elem())
}

func stableTypeName(typeToken reflect.Type) string {
	if typeToken.Name() != "" {
		if typeToken.PkgPath() != "" {
			return typeToken.PkgPath() + "." + typeToken.Name()
		}
		return typeToken.Name()
	}
	switch typeToken.Kind() {
	case reflect.Pointer:
		return "*" + stableTypeName(typeToken.Elem())
	case reflect.Slice:
		return "[]" + stableTypeName(typeToken.Elem())
	case reflect.Array:
		return fmt.Sprintf("[%d]%s", typeToken.Len(), stableTypeName(typeToken.Elem()))
	case reflect.Map:
		return "map[" + stableTypeName(typeToken.Key()) + "]" + stableTypeName(typeToken.Elem())
	case reflect.Chan:
		return typeToken.ChanDir().String() + " " + stableTypeName(typeToken.Elem())
	case reflect.Func:
		var name strings.Builder
		name.WriteString("func(")
		for index := 0; index < typeToken.NumIn(); index++ {
			if index > 0 { name.WriteByte(',') }
			name.WriteString(stableTypeName(typeToken.In(index)))
		}
		name.WriteByte(')')
		if typeToken.NumOut() == 1 {
			name.WriteByte(' ')
			name.WriteString(stableTypeName(typeToken.Out(0)))
		} else if typeToken.NumOut() > 1 {
			name.WriteString(" (")
			for index := 0; index < typeToken.NumOut(); index++ {
				if index > 0 { name.WriteByte(',') }
				name.WriteString(stableTypeName(typeToken.Out(index)))
			}
			name.WriteByte(')')
		}
		return name.String()
	}
	return typeToken.String()
}

func Specialize(member Member, callable, receiver string, receiverTypeArgumentCount int, typeArguments ...string) Member {
	if receiver != "" {
		prefix := member.Module + "." + receiver
		receiverArguments := typeArguments[:receiverTypeArgumentCount]
		member.Method = prefix + "[" + strings.Join(receiverArguments, ",") + "]" + strings.TrimPrefix(member.Method, prefix)
	} else {
		prefix := member.Module + "." + callable
		member.Method = prefix + "[" + strings.Join(typeArguments, ",") + "]" + strings.TrimPrefix(member.Method, prefix)
	}
	member.Status = "Patched"
	member.SkipReason = ""
	member.Detail = ""
	return member
}

func ensureMemberLocked(member Member) bool {
	module := state.modules[member.Module]
	if module == nil {
		module = &moduleState{members: make(map[string]Member)}
		state.modules[member.Module] = module
	}
	if _, exists := module.members[member.Method]; exists {
		return false
	}
	module.members[member.Method] = member
	return true
}

func Enter(parent *Frame, member Member, args []any) *Frame {
	testID := "(no-test)"
	parentCallID := uint64(0)
	depth := 0
	if parent != nil {
		testID = parent.testID
		parentCallID = parent.callID
		depth = parent.depth + 1
	}
	return enter(parentCallID, depth, testID, member, args)
}

func OpenTest(testID string, member Member, args []any) *Frame {
	if testID == "" {
		testID = "(no-test)"
	}
	return enter(0, 0, testID, member, args)
}

func enter(parentCallID uint64, depth int, testID string, member Member, args []any) *Frame {
	state.Lock()
	if ensureMemberLocked(member) {
		writeManifestLocked()
	}
	callID := state.nextCallID
	state.nextCallID++
	key := ordinalKey{testID: testID, method: member.Method}
	ordinal := state.ordinals[key]
	state.ordinals[key] = ordinal + 1
	state.Unlock()
	argsCapture := captureValue(args, member, member.ParameterNames)
	return &Frame{
		member: member,
		testID: testID,
		callID: callID,
		parentCallID: parentCallID,
		depth: depth,
		ordinal: ordinal,
		args: argsCapture,
	}
}

func Exit(frame *Frame, results []any, recovered any) {
	if frame == nil {
		return
	}
	frame.exit.Do(func() {
		event := map[string]any{
			"testId": frame.testID,
			"methodFullName": frame.member.Method,
			"filePathResolution": frame.member.SourceResolution,
			"line": frame.member.Line,
			"callDepth": frame.depth,
			"callId": frame.callID,
			"ordinal": frame.ordinal,
			"threadId": 0,
		}
		if frame.member.File != "" {
			event["filePath"] = frame.member.File
		}
		if frame.parentCallID != 0 {
			event["parentCallId"] = frame.parentCallID
		}
		if frame.args != nil {
			event["argsDigest"] = frame.args.digest
			event["argsRendered"] = frame.args.rendered
		}
		if recovered != nil {
			event["exceptionType"] = fmt.Sprintf("%T", recovered)
		} else if len(results) > 0 {
			if captured := captureValue(resultValues(results), frame.member, nil); captured != nil {
				event["returnDigest"] = captured.digest
				event["returnRendered"] = captured.rendered
			}
		}
		if frame.member.IsHarness {
			event["isHarness"] = true
		}
		line, err := json.Marshal(event)
		if err != nil {
			panic(fmt.Errorf("behaviordiff: encode event: %w", err))
		}
		state.Lock()
		defer state.Unlock()
		if state.trace == nil {
			return
		}
		state.enqueued++
		if _, err := state.trace.Write(append(line, '\n')); err != nil {
			panic(fmt.Errorf("behaviordiff: write event: %w", err))
		}
		state.written++
		if module := state.modules[frame.member.Module]; module != nil {
			module.tracedCalls++
		}
		writeManifestLocked()
	})
}

func Go(parent *Frame, fn func(*Frame)) {
	captured := parent
	go func() {
		fn(captured)
	}()
}

func captureValue(value any, member Member, argumentNames []string) *capture {
	result := DigestWithOptions(value, Options{
		Redact: true,
		SensitiveNames: sensitiveNames,
		DigestOnlyTypes: digestOnlyTypes,
		ArgumentNames: argumentNames,
	})
	state.Lock()
	state.digest.Values += result.Counters.Values
	state.digest.DepthLimited += result.Counters.DepthLimited
	state.digest.EntryLimited += result.Counters.EntryLimited
	state.digest.Blocklisted += result.Counters.Blocklisted
	state.digest.MapTies += result.Counters.MapTies
	state.digest.Errored += result.Counters.Errored
	state.digest.RenderedTruncated += result.Counters.RenderedTruncated
	state.digest.Unexported += result.Counters.Unexported
	state.Unlock()
	rendered := result.Canonical
	if digestOnlyPath(member.File) {
		rendered = "<redacted>"
	}
	return &capture{digest: "sha256:" + result.SHA256, rendered: rendered}
}

func configuredList(name string) []string {
	var values []string
	for _, value := range strings.FieldsFunc(os.Getenv(name), func(character rune) bool { return character == ';' || character == ',' }) {
		if trimmed := strings.TrimSpace(value); trimmed != "" {
			values = append(values, trimmed)
		}
	}
	return values
}

func digestOnlyPath(path string) bool {
	normalized := strings.Trim(strings.ReplaceAll(path, "\\", "/"), "/")
	for _, configured := range digestOnlyPaths {
		prefix := strings.Trim(strings.ReplaceAll(configured, "\\", "/"), "/")
		if normalized == prefix || strings.HasPrefix(normalized, prefix+"/") ||
			strings.HasSuffix(normalized, "/"+prefix) || strings.Contains(normalized, "/"+prefix+"/") {
			return true
		}
	}
	return false
}

func resultValues(results []any) any {
	values := make([]any, len(results))
	for index, result := range results {
		value := reflect.ValueOf(result)
		if value.IsValid() && value.Kind() == reflect.Pointer && !value.IsNil() {
			values[index] = value.Elem().Interface()
		} else {
			values[index] = result
		}
	}
	if len(values) == 1 {
		return values[0]
	}
	return values
}

func decorate(configured, suffix string) string {
	absolute, err := filepath.Abs(configured)
	if err != nil {
		panic(fmt.Errorf("behaviordiff: resolve trace path: %w", err))
	}
	extension := filepath.Ext(absolute)
	base := strings.TrimSuffix(filepath.Base(absolute), extension)
	name := fmt.Sprintf("%s.%d%s%s", base, os.Getpid(), suffix, extension)
	return filepath.Join(filepath.Dir(absolute), name)
}

func writeManifestLocked() {
	if state.manifestPath == "" {
		return
	}
	records := make([]map[string]any, 0)
	records = append(records, map[string]any{"kind": "run", "schema": schema, "language": language})
	moduleNames := make([]string, 0, len(state.modules))
	for name := range state.modules {
		moduleNames = append(moduleNames, name)
	}
	sort.Strings(moduleNames)
	for _, name := range moduleNames {
		module := state.modules[name]
		members := make([]Member, 0, len(module.members))
		for _, member := range module.members {
			members = append(members, member)
		}
		sort.Slice(members, func(left, right int) bool { return members[left].Method < members[right].Method })
		patched := 0
		exact := 0
		for _, member := range members {
			if member.Status == "Patched" {
				patched++
				if member.SourceResolution == "debugInfo" {
					exact++
				}
			}
		}
		skipped := len(members) - patched
		exactPercent := 100
		sourceRule := "notApplicable"
		if patched > 0 {
			exactPercent = exact * 100 / patched
			sourceRule = "ratio"
		}
		records = append(records, map[string]any{
			"kind": "assembly", "assembly": name, "discovery": "GoAstRewrite",
			"scanned": true, "instrumented": patched > 0, "patchedMembers": patched,
			"discoveredMembers": len(members), "skippedMembers": skipped, "patchFailedMembers": 0,
			"queuedAtMs": 0, "patchedAtMs": 0, "tracedCalls": module.tracedCalls,
			"membersWithExactSource": exact, "exactSourcePercent": exactPercent, "sourceRule": sourceRule,
		})
		for _, member := range members {
			record := map[string]any{
				"kind": "member", "assembly": name, "method": member.Method,
				"status": member.Status, "returnKind": member.ReturnKind,
				"sourceResolution": member.SourceResolution,
			}
			if member.SkipReason != "" { record["skipReason"] = member.SkipReason }
			if member.Detail != "" { record["detail"] = member.Detail }
			if member.IsTestRoot { record["isTestRoot"] = true }
			records = append(records, record)
		}
	}
	records = append(records, map[string]any{
		"kind": "digest", "valuesDigested": state.digest.Values,
		"depthLimited": state.digest.DepthLimited, "blocklisted": state.digest.Blocklisted,
		"errored": state.digest.Errored, "renderedTruncated": state.digest.RenderedTruncated,
		"unreadableFields": state.digest.Unexported, "ambiguousMapEntries": state.digest.MapTies,
	})
	records = append(records, map[string]any{
		"kind": "writer", "enqueued": state.enqueued, "written": state.written,
		"dropped": 0, "capacity": writerCapacity,
	})
	var output strings.Builder
	for _, record := range records {
		line, err := json.Marshal(record)
		if err != nil { panic(fmt.Errorf("behaviordiff: encode manifest: %w", err)) }
		output.Write(line)
		output.WriteByte('\n')
	}
	temporary := state.manifestPath + ".tmp"
	if err := os.WriteFile(temporary, []byte(output.String()), 0644); err != nil {
		panic(fmt.Errorf("behaviordiff: write manifest: %w", err))
	}
	if err := os.Rename(temporary, state.manifestPath); err != nil {
		_ = os.Remove(state.manifestPath)
		if retryErr := os.Rename(temporary, state.manifestPath); retryErr != nil {
			panic(fmt.Errorf("behaviordiff: replace manifest: %w", retryErr))
		}
	}
}
`

func writeRuntime(out string) error {
	directory := filepath.Join(out, "internal", "behaviordiffrt")
	if err := os.MkdirAll(directory, 0o755); err != nil {
		return fmt.Errorf("create runtime package: %w", err)
	}
	formattedRuntime, err := format.Source([]byte(runtimeSource))
	if err != nil {
		return fmt.Errorf("format runtime source: %w", err)
	}
	if err := os.WriteFile(filepath.Join(directory, "runtime.go"), formattedRuntime, 0o644); err != nil {
		return fmt.Errorf("write runtime package: %w", err)
	}
	_, sourceFile, _, ok := runtime.Caller(0)
	if !ok {
		return fmt.Errorf("locate canonicalizer source")
	}
	canonicalPath := filepath.Join(filepath.Dir(sourceFile), "..", "runtime", "canonical", "canonical.go")
	canonical, err := os.ReadFile(canonicalPath)
	if err != nil {
		return fmt.Errorf("read canonicalizer source: %w", err)
	}
	generated := strings.NewReplacer(
		"package canonical", "package behaviordiffrt",
		"newState", "newCanonicalState",
		"type state struct", "type canonicalState struct",
		"*state", "*canonicalState",
		"&state{", "&canonicalState{",
	).Replace(string(canonical))
	formattedCanonical, err := format.Source([]byte(generated))
	if err != nil {
		return fmt.Errorf("format canonicalizer source: %w", err)
	}
	if err := os.WriteFile(filepath.Join(directory, "canonical.go"), formattedCanonical, 0o644); err != nil {
		return fmt.Errorf("write canonicalizer source: %w", err)
	}
	return nil
}

File: src/RealDiff.Python/realdiff_python/runtime.py
from __future__ import annotations

import atexit
import json
import os
import threading
from collections import defaultdict
from dataclasses import dataclass
from pathlib import Path
from types import CodeType, FrameType

from .canonical import Canonicalizer, DigestStatistics, MISSING
from .monitor import Scope


@dataclass
class CallState:
    call_id: int
    parent_call_id: int | None
    depth: int
    ordinal: int
    method: str
    file_path: str
    line: int
    thread_id: int
    args_digest: str
    args_rendered: str
    test_id: str = "(no-test)"
    pending_exception: BaseException | None = None
    suspended: bool = False


@dataclass(frozen=True)
class Member:
    module: str
    method: str
    file_path: str
    line: int
    status: str
    return_kind: str
    skip_reason: str | None = None
    detail: str | None = None


class Runtime:
    def __init__(self, trace_path: Path, scope: Scope):
        self.trace_path = trace_path
        self.manifest_path = trace_path.with_name(trace_path.stem + ".manifest.ndjson")
        self.scope = scope
        self._lock = threading.RLock()
        self._next_call_id = 1
        self._states: dict[int, CallState] = {}
        self._stacks: dict[int, list[int]] = defaultdict(list)
        self._ordinals: dict[tuple[str, str], int] = defaultdict(int)
        self._members: dict[str, Member] = {}
        self._module_calls: dict[str, int] = defaultdict(int)
        self._digest_statistics = DigestStatistics()
        self._canonicalizer = Canonicalizer(statistics=self._digest_statistics)
        self._written = 0
        self._closed = False
        trace_path.parent.mkdir(parents=True, exist_ok=True)
        self._stream = trace_path.open("w", encoding="utf-8", newline="\n")
        atexit.register(self.close)

    @classmethod
    def from_environment(cls) -> "Runtime":
        trace = os.environ.get("REALDIFF_TRACE")
        if not trace:
            raise RuntimeError("REALDIFF_TRACE is required for Python tracing")
        return cls(Path(trace).resolve(), Scope.from_environment())

    def on_event(self, event: str, code: CodeType, value: object | None, frame: FrameType) -> None:
        with self._lock:
            if event == "start":
                self._start(code, frame)
            elif event == "resume":
                self._resume(frame)
            elif event == "yield":
                self._suspend(frame)
            elif event == "raise":
                self._raised(frame, value)
            elif event == "return":
                self._complete(frame, return_value=value)
            elif event == "unwind":
                self._complete(frame, exception=value)

    def _start(self, code: CodeType, frame: FrameType) -> None:
        file_path = self.scope.relative_path(code)
        if file_path is None:
            return
        module = _module_name(file_path)
        method = _method_name(file_path, code)
        if code.co_name == "<module>":
            self._members.setdefault(
                method,
                Member(module, method, file_path, code.co_firstlineno, "Skipped", "Module", "UnsupportedShape", "Python: ModuleBody"),
            )
            return

        thread_id = threading.get_native_id()
        stack = self._stacks[thread_id]
        call_id = self._next_call_id
        self._next_call_id += 1
        test_id = "(no-test)"
        ordinal_key = (test_id, method)
        ordinal = self._ordinals[ordinal_key]
        self._ordinals[ordinal_key] += 1
        state = CallState(
            call_id=call_id,
            parent_call_id=stack[-1] if stack else None,
            depth=len(stack),
            ordinal=ordinal,
            method=method,
            file_path=file_path,
            line=max(1, code.co_firstlineno),
            thread_id=thread_id,
            args_digest="",
            args_rendered="",
        )
        arguments = _capture_arguments(code, frame)
        canonical_arguments = self._canonicalizer.digest(arguments, source_path=file_path)
        state.args_digest = canonical_arguments.digest
        state.args_rendered = canonical_arguments.rendered
        self._states[id(frame)] = state
        stack.append(call_id)
        self._members.setdefault(
            method,
            Member(module, method, file_path, state.line, "Patched", _return_kind(code)),
        )

    def _resume(self, frame: FrameType) -> None:
        state = self._states.get(id(frame))
        if state is None or not state.suspended:
            return
        stack = self._stacks[state.thread_id]
        state.parent_call_id = stack[-1] if stack else None
        state.depth = len(stack)
        stack.append(state.call_id)
        state.suspended = False

    def _suspend(self, frame: FrameType) -> None:
        state = self._states.get(id(frame))
        if state is None:
            return
        self._pop_stack(state)
        state.suspended = True

    def _raised(self, frame: FrameType, value: object | None) -> None:
        state = self._states.get(id(frame))
        if state is not None and value is not None:
            state.pending_exception = value

    def _complete(
        self,
        frame: FrameType,
        return_value: object | None = None,
        exception: object | None = None,
    ) -> None:
        state = self._states.pop(id(frame), None)
        if state is None:
            return
        self._pop_stack(state)
        record: dict[str, object] = {
            "testId": state.test_id,
            "methodFullName": state.method,
            "filePath": state.file_path,
            "filePathResolution": "debugInfo",
            "line": state.line,
            "callDepth": state.depth,
            "callId": state.call_id,
            "ordinal": state.ordinal,
            "argsDigest": state.args_digest,
            "argsRendered": state.args_rendered,
            "threadId": state.thread_id,
        }
        if state.parent_call_id is not None:
            record["parentCallId"] = state.parent_call_id
        if exception is not None:
            record["exceptionType"] = _exception_name(exception)
        else:
            canonical_return = self._canonicalizer.digest(return_value, source_path=state.file_path)
            record["returnDigest"] = canonical_return.digest
            record["returnRendered"] = canonical_return.rendered
        self._write(record)
        self._module_calls[_module_name(state.file_path)] += 1

    def _pop_stack(self, state: CallState) -> None:
        stack = self._stacks[state.thread_id]
        if stack and stack[-1] == state.call_id:
            stack.pop()

    def _write(self, record: dict[str, object]) -> None:
        self._stream.write(json.dumps(record, sort_keys=True, separators=(",", ":")) + "\n")
        self._stream.flush()
        self._written += 1

    def close(self) -> None:
        with self._lock:
            if self._closed:
                return
            self._closed = True
            self._stream.close()
            self._write_manifest()

    def _write_manifest(self) -> None:
        by_module: dict[str, list[Member]] = defaultdict(list)
        for member in self._members.values():
            by_module[member.module].append(member)
        with self.manifest_path.open("w", encoding="utf-8", newline="\n") as stream:
            _write_line(stream, {"kind": "run", "schema": "realdiff.trace/1", "language": "python"})
            for module in sorted(by_module):
                members = sorted(by_module[module], key=lambda item: item.method)
                patched = sum(member.status == "Patched" for member in members)
                _write_line(
                    stream,
                    {
                        "kind": "assembly",
                        "assembly": module,
                        "discovery": "SysMonitoring",
                        "scanned": True,
                        "instrumented": patched > 0,
                        "patchedMembers": patched,
                        "discoveredMembers": len(members),
                        "skippedMembers": len(members) - patched,
                        "patchFailedMembers": 0,
                        "queuedAtMs": 0,
                        "patchedAtMs": 0,
                        "tracedCalls": self._module_calls[module],
                        "membersWithExactSource": patched,
                        "exactSourcePercent": 100,
                        "sourceRule": "ratio",
                    },
                )
                for member in members:
                    record: dict[str, object] = {
                        "kind": "member",
                        "assembly": member.module,
                        "method": member.method,
                        "filePath": member.file_path,
                        "line": member.line,
                        "status": member.status,
                        "returnKind": member.return_kind,
                        "sourceResolution": "debugInfo",
                    }
                    if member.skip_reason is not None:
                        record["skipReason"] = member.skip_reason
                    if member.detail is not None:
                        record["detail"] = member.detail
                    _write_line(stream, record)
            _write_line(
                stream,
                {
                    "kind": "digest",
                    "valuesDigested": self._digest_statistics.values_digested,
                    "depthLimited": self._digest_statistics.depth_limited,
                    "blocklisted": self._digest_statistics.blocklisted,
                    "errored": self._digest_statistics.errored,
                    "renderedTruncated": self._digest_statistics.rendered_truncated,
                    "unreadableFields": self._digest_statistics.unreadable_fields,
                    "ambiguousMapEntries": self._digest_statistics.ambiguous_map_entries,
                },
            )
            _write_line(
                stream,
                {"kind": "writer", "enqueued": self._written, "written": self._written, "dropped": 0, "capacity": 1},
            )


def _write_line(stream, record: dict[str, object]) -> None:
    stream.write(json.dumps(record, sort_keys=True, separators=(",", ":")) + "\n")


def _module_name(file_path: str) -> str:
    path = Path(file_path)
    return ".".join(path.with_suffix("").parts)


def _method_name(file_path: str, code: CodeType) -> str:
    return f"{_module_name(file_path)}::{code.co_qualname}"


def _return_kind(code: CodeType) -> str:
    if code.co_flags & 0x200:
        return "AsyncGenerator"
    if code.co_flags & 0x80:
        return "Coroutine"
    if code.co_flags & 0x20:
        return "Generator"
    return "Sync"


def _exception_name(exception: BaseException) -> str:
    type_ = type(exception)
    return f"{type_.__module__}.{type_.__qualname__}"


def _capture_arguments(code: CodeType, frame: FrameType) -> dict[str, object]:
    count = code.co_argcount + code.co_kwonlyargcount
    names = list(code.co_varnames[:count])
    next_index = count
    if code.co_flags & 0x04:
        names.append(code.co_varnames[next_index])
        next_index += 1
    if code.co_flags & 0x08:
        names.append(code.co_varnames[next_index])
    locals_ = frame.f_locals
    return {name: dict.get(locals_, name, MISSING) for name in names}